projects/core/src/routing/event/routing-event.builder.ts

```diff
@@ -0,0 +1,164 @@
+  CategoryPageVisited,
+  HomePageVisited,
+  SearchPageVisited,
+  OrderConfirmationPageVisited,
+  PageVisited,
+  ProductDetailsPageVisited,
+} from './routing.events';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class RoutingEventBuilder {
+  constructor(
+    protected actions: ActionsSubject,
+    protected eventService: EventService,
+    protected productService: ProductService,
+    protected productSearchService: ProductSearchService,
+    protected semanticPathService: SemanticPathService
+  ) {
+    this.register();
+  }
+
+  protected register(): void {
+    this.eventService.register(
+      SearchPageVisited,
+      this.searchResultPageVisited()
+    );
+    this.eventService.register(
+      ProductDetailsPageVisited,
+      this.buildProductDetailsPageVisitedEvent()
+    );
+    this.eventService.register(
+      CategoryPageVisited,
+      this.buildCategoryPageVisitedEvent()
+    );
+    this.eventService.register(
+      HomePageVisited,
+      this.buildHomePageVisitedEvent()
+    );
+    this.eventService.register(CartPageVisited, this.buildCartVisitedEvent());
+    this.eventService.register(PageVisited, this.buildPageVisitedEvent());
+    this.eventService.register(
+      OrderConfirmationPageVisited,
+      this.orderConfirmationPageVisitedEvent()
+    );
+  }
+
+  protected buildPageVisitedEvent(): Observable<PageVisited> {
+    return this.getNavigatedEvent().pipe(
+      map((state) => createFrom(PageVisited, state))
+    );
+  }
+
+  protected buildHomePageVisitedEvent(): Observable<HomePageVisited> {
+    return this.getCurrentPageContextFor('home').pipe(
+      map((pageContext) => createFrom(HomePageVisited, pageContext))
+    );
+  }
+
+  protected buildProductDetailsPageVisitedEvent(): Observable<
+    ProductDetailsPageVisited
+  > {
+    return this.getCurrentPageContextFor('product').pipe(
+      map((context) => context.id),
+      switchMap((productId) => {
+        return this.productService.get(productId).pipe(
+          filter(Boolean),
+          map((product) => {
+            return createFrom(ProductDetailsPageVisited, product);
+          })
+        );
+      })
+    );
+  }
+
+  protected buildCartVisitedEvent(): Observable<CartPageVisited> {
+    return this.getCurrentPageContextFor('cart').pipe(
+      map((pageContext) => createFrom(CartPageVisited, pageContext))
+    );
+  }
+
+  protected orderConfirmationPageVisitedEvent(): Observable<
+    OrderConfirmationPageVisited
+  > {
+    return this.getCurrentPageContextFor('orderConfirmation').pipe(
+      map((pageContext) =>
+        createFrom(OrderConfirmationPageVisited, pageContext)
+      )
+    );
+  }
+
+  //FIXME: see https://github.com/SAP/spartacus/issues/7991
+  protected buildCategoryPageVisitedEvent(): Observable<CategoryPageVisited> {
+    return this.productSearchService.getResults().pipe(
+      filter((searchResults) => {
+        return (
+          searchResults.breadcrumbs && searchResults.breadcrumbs.length > 0
```

Comment:
Bagatelle: What do you think about using shortcut - optional chaining `searchResults.breadcrumbs?.length`?

Reply:
It appears this has been moved to another PR